big reshuffle:

ara4n opened the pull request:
* try to make 00_basis slightly more accessible and less terse
 * move CS API before Events, as it's meaningless to dive into specific events if you haven't been introduced to how the CS API works
 * Try to improve the beginning of the CS API by merging back in some of kegan's original v1 content (which has subsequently been deleted somewhere along the line?)
 * document m.presence event consistently
 * move typing from drafts to spec
 * move push stuff from drafts to spec
...and lots of other stuff

move push & typing from drafts into main spec.

Kegsay commented on specification/00_basis.rst:
```diff
@@ -355,38 +374,31 @@ To be added to this list, the user being added must be invited by the list owner
 
-.. TODO-spec
-  - Metadata extensibility
+..TODO
```
Should be ``.. TODO`` (missing the space), which is required in order for it to be a comment.

Kegsay commented on specification/00_basis.rst:
```diff
@@ -355,38 +374,31 @@ To be added to this list, the user being added must be invited by the list owner
+client preferences, or server configuration settings which lack any other
+dedicated API.  The API is symmetrical to managing Profile data.
 
+..TODO
```
Space missing.

Kegsay commented on specification/10_client_server_api.rst:
```diff
@@ -1,377 +1,437 @@
+    pagination using ``from`` and ``to`` query parameters, coupled with the
+    ``start`` and ``end`` tokens from an |initialSync|_ API.
+    
+    XXX: Is this accurate? Doesn't it return all events - not just m.room.message/member?
```
This is not accurate: it returns all events.

As an aside: please use RST comments for meta-text so they won't show up when rendered as HTML, e.g.
```
visible stuff

.. XXX: Is this accurate? Doesn't it return all events - not just m.room.message/member?

more visible stuff
```

Kegsay commented on drafts/general_api.rst:
```diff
@@ -810,6 +810,10 @@ Notes:
 Presence API ``[Draft]``
 ------------------------
+
+FIXME: this seems to be ignoring activity timers entirely, which were present on
```
Whilst I appreciate this is only a draft, please put this as a RST comment to make it easier when moving over to official spec docs.

Kegsay commented on scripts/gendoc.py:
```diff
@@ -101,7 +101,8 @@ def prepare_env():
 def cleanup_env():
-    shutil.rmtree("./tmp")
+    pass
+    #shutil.rmtree("./tmp")
```
Explanation here?

Kegsay commented on specification/00_basis.rst:
```diff
@@ -13,11 +13,14 @@ Table of Contents
+Matrix is a set of open APIs for open-federated Instant Messaging, VoIP and
+Internet of Things communication, designed to create and support a new global
+real-time communication ecosystem. The intention is to provide an open
+decentralised pubsub fabric for the internet for securely persisting and
```
Please don't use the word ``fabric`` here. As a techie, I don't get what that word is supposed to mean (as opposed to later on when you say ``JSON`` and ``publishing/subscribing`` which are common terms). To the layman, I don't think ``fabric`` makes it any easier either, so who is the target audience? I don't think it can be used as a metaphor to help understand the concept either. Given this document is aimed at a technical audience, I don't think this is the kind of wording we should be using, *especially* in the first paragraph in the introduction (hence why I'm bringing this up).

Kegsay commented on specification/00_basis.rst:
```diff
@@ -13,11 +13,14 @@ Table of Contents
-communication ecosystem on the internet. This specification is the ongoing
-result of standardising the APIs used by the various components of the Matrix
-ecosystem to communicate with one another.
+Matrix is a set of open APIs for open-federated Instant Messaging, VoIP and
```
It would probably be polite for us to spell out acronyms like ``VoIP`` when we first use them.

Kegsay commented on specification/00_basis.rst:
```diff
@@ -97,8 +100,34 @@ Overview
-not communicate with each other directly.
+Matrix defines APIs for synchronising extensible JSON objects known as
+`events` between compatible clients, servers and services. Clients are
+typically messaging/VoIP applications or IoT devices/hubs and communicate by
```
We should spell out the acronym ``IoT`` here.

Kegsay commented on specification/00_basis.rst:
```diff
@@ -97,8 +100,34 @@ Overview
-Clients transmit data to other clients through home servers (HSes). Clients do
-not communicate with each other directly.
+Matrix defines APIs for synchronising extensible JSON objects known as
+`events` between compatible clients, servers and services. Clients are
```
Syntax: Single `` ` `` will be rendered as *italics* in RST, not ``block`` which may be your intention here? If you intended to use italics, I would prefer ``block`` simply because it makes it clearer that it is a "special term" rather than an emphasised word.

Kegsay commented on specification/00_basis.rst:
```diff
@@ -97,8 +100,34 @@ Overview
+single homeserver has control or ownership over a given room*. Homeservers
+model communication history as a partially ordered graph of events known as
+the room's `event graph`, which is synchronised with eventual consistency
+between the participating servers using the Server-Server API. Matrix optimises
```
Consistency: should probably highlight ``Server-Server API`` in the same way you did in the CS API on ``:107``

Kegsay commented on specification/00_basis.rst:
```diff
@@ -107,21 +136,22 @@ not communicate with each other directly.
        |                  |---------( HTTPS )--------->|                  |
        |   Home Server    |                            |   Home Server    |
        |                  |<--------( HTTPS )----------|                  |
-       +------------------+        Federation          +------------------+
```
Any particular reason for phasing out the word ``Federation``? I think your replacement text is more descriptive, but due to the proliferance of the word elsewhere in the code and spec, I think highlighting it at this point makes sense.

Kegsay commented on specification/00_basis.rst:
```diff
@@ -131,19 +161,14 @@ this user. They are case-insensitive.
 
-A "Home Server" is a server which provides C-S APIs and has the ability to
-federate with other HSes.  It is typically responsible for multiple clients.
-"Federation" is the term used to describe the sharing of data between two or
```
As I said on ``:116``, we should really be saying what Federation is, given this is the "Architecture" section.

Kegsay commented on specification/00_basis.rst:
```diff
@@ -200,7 +229,7 @@ the room ``!qporfwt:matrix.org``::
        +------------------+                          +------------------+
                |                                                 ^
-               |                                                 |
+               |         [HTTP POST]                             |
```
Nitpick: I don't think this is a ``POST`` in the Federation API, I think it's a ``PUT``.

Kegsay commented on specification/00_basis.rst:
```diff
@@ -306,47 +335,37 @@ the Matrix ecosystem. However, without one clients will not be able to look up
+  The last_active specifics should be moved to the detailed presence event section
+  
+Last activity is tracked by the server maintaining a timestamp of the last time
+it saw a pro-active event from the user; either sending a message to a room,
```
We need to be more specific than this. "Pro-active" as a term isn't clear enough to say when a user is treated as going back online. You mention "sending a message" but we need to say what a "message" is: strictly ``m.room.message``? Confusion here could lead to server implementations treating last activity time differently.

Kegsay commented on specification/00_basis.rst:
```diff
@@ -355,38 +374,31 @@ To be added to this list, the user being added must be invited by the list owner
-Profiles
-++++++++
+Profile data is typed using namespaced keys for interoperability, much like
+events - e.g. ``m.profile.display_name``.
```
Is this true? It isn't in v1, and I was unaware this was specced in v2.

Kegsay commented on specification/00_basis.rst:
```diff
@@ -355,38 +374,31 @@ To be added to this list, the user being added must be invited by the list owner
 
+..TODO
+  Would it really be overengineered to use the same API for both profile &
+  private user data, but with different ACLs?
```
How servers wish to implement it is up to them, we just need to say that it is "private" and shouldn't be accessed by anyone other than the user who ``PUT`` it, surely.

Kegsay commented on specification/00_basis.rst:
```diff
@@ -430,6 +442,9 @@ For example, if there was a custom namespace ``com.mydomain.here``, and a
 error, but the keys ``error`` and ``errcode`` MUST always be present.
 
+..TODO
+  Why the weird mix of underscore and dots?
```
IIRC this was to make it easier to determine the boundary between the namespacing and the error "last path segment", because you can just do ``"COM.STUFF.NOT_FOUND".split(".")[-1]``. The main benefit for this would be in custom app error handling.

Kegsay commented on specification/10_client_server_api.rst:
```diff
@@ -1,377 +1,437 @@
+    integer - An integer representing the maximum number of items to 
+    return.
+
+'START' and 'END' are magic token values which specify the start and end of the 
```
I'm not sure if ``START`` and ``END`` are actually used in v1 anymore. The equivalent operation is obtained by omitting the token iirc.

Kegsay commented on specification/10_client_server_api.rst:
```diff
@@ -1,377 +1,437 @@
+last request left off.
 
-To respond to this type, reply with::
+All events must be deduplicated based on their event ID (TODO: is this actually
```
Please use a RST comment for this. Unsure if this is a hard requirement for v2 (given I don't know enough federation magic).

Kegsay commented on specification/10_client_server_api.rst:
```diff
@@ -1,377 +1,437 @@
 
-To respond to this type, reply with::
+:Message events:
+  These are events which describe transient "once-off" activity in a room:
```
``s/once/one/``

ara4n replied:
i believe this is to personal taste - personally "once-off" feels much better than "one-off" here.

Kegsay commented on specification/10_client_server_api.rst:
```diff
@@ -1,377 +1,437 @@
+  Response format:
+    `` { TODO-doc } ``
+  Example:
+    TODO-doc
```
Please add RST comments to these two ``TODO-doc`` bits.

ara4n replied:
Given almost all the examples & responses here are non-commented TODOs, i'm not going to special case these unless I'm missing something.

Kegsay commented on specification/10_client_server_api.rst:
```diff
@@ -1,377 +1,437 @@
-  GET matrix/client/api/v1/login/fallback
+Currently, only room admins can redact events by sending a ``m.room.redaction``
+event, but server admins also need to be able to redact events by a similar
+mechanism.
```
This sounds like it isn't implemented for server admins. If it isn't, it should be bugged and this should be marked as a ``TODO``, rather than casually mentioning a missing feature in the spec.

Kegsay commented on specification/10_client_server_api.rst:
```diff
@@ -1,377 +1,437 @@
+ - ``m.room.member`` allows key ``membership``
+ - ``m.room.create`` allows key ``creator``
+ - ``m.room.join_rules`` allows key ``join_rule``
+ - ``m.room.power_levels`` allows keys that are user ids or ``default``
```
This needs to be updated in light of the power level key changes done a while back, which now look like:
```
{
  "ban": 50,
  "events": {
    "m.room.name": 100,
    "m.room.power_levels": 100
  },
  "events_default": 0,
  "kick": 50,
  "redact": 50,
  "state_default": 50,
  "users": {
    "@example:localhost": 100
  },
  "users_default": 0
}
```

Kegsay commented on specification/20_events.rst:
```diff
@@ -447,6 +448,48 @@ outlined below:
 .. TODO-spec
     Make the definitions "inherit" from FileInfo where necessary...
 
+Presence Events (v1)
```
Given the whole CS spec is v1, it seems odd to specify this here.

Kegsay commented on specification/20_events.rst:
```diff
@@ -447,6 +448,48 @@ outlined below:
+  Type:
+    Presence event
+  JSON format::
+    { "displayname": "utf-8 string",
```
Formatting: I would prefer this to be on a new line to match the form:
```
{
  key: value,
  key: value
  ...
}
```
which I think looks neater and more symmetric than:
```
{ key: value,
  key: value,
  ... }
```
I don't think we need to be optimising for line counts.

Kegsay commented on specification/42_push_overview.rst:
```diff
@@ -1,5 +1,5 @@
-Push Notifications
-==================
+Push Notifications Overview
+===========================
```
``==`` is the top-level title heading. We should really have the title headings be:
```
Push Notifications
  Overview
  HTTP API
```
rather than:
```
Push Notifications Overview
Push Notifications HTTP API
```

ara4n commented:
I think I've addressed all of the feedback above.  Kegan: beyond the minutiae, any thoughts on whether this restructuring actually constitutes an improvement?
